File: python/courtlistener/utils/validators.py
"""
Validation utilities for the CourtListener SDK.
"""

import re
from typing import Optional, Union
from datetime import date, datetime
from ..exceptions import ValidationError


def validate_date(date_str: str) -> bool:
    """
    Validate date string format (YYYY-MM-DD).
    
    Args:
        date_str: Date string to validate
    
    Returns:
        True if valid, raises ValidationError if invalid
    
    Raises:
        ValidationError: If date format is invalid
    """
    if not date_str:
        raise ValidationError("Date string cannot be empty")
    
    # Check format YYYY-MM-DD
    pattern = r'^\d{4}-\d{2}-\d{2}$'
    if not re.match(pattern, date_str):
        raise ValidationError(f"Invalid date format: {date_str}. Expected YYYY-MM-DD")
    
    try:
        # Validate that it's a real date
        datetime.strptime(date_str, '%Y-%m-%d')
        return True
    except ValueError as e:
        raise ValidationError(f"Invalid date: {date_str}. {str(e)}")


def validate_citation(citation: str) -> bool:
    """
    Validate legal citation format.
    
    Supports common citation formats including:
    - SCOTUS: "576 U.S. 644" or "576 US 644"
    - Federal Reporter: "123 F.3d 456" or "123 F3d 456" or "123 F. 3d 456"
    - Federal Supplement: "123 F.Supp. 456" or "123 F Supp 456"
    - Federal Appendix: "123 F.App'x 456"
    - State citations: "123 Cal. 3d 456" or "123 N.Y. 456"
    
    Args:
        citation: Citation string to validate
    
    Returns:
        True if valid, raises ValidationError if invalid
    
    Raises:
        ValidationError: If citation format is invalid
    """
    if not citation:
        raise ValidationError("Citation cannot be empty")
    
    # Normalize whitespace
    citation_clean = re.sub(r'\s+', ' ', citation.strip())
    
    # Comprehensive citation patterns
    patterns = [
        # SCOTUS: 576 U.S. 644 or 576 US 644
        r'^\d+\s+U\.?S\.?\s+\d+$',
        # Federal Reporter: 123 F.3d 456, 123 F3d 456, 123 F. 3d 456
        r'^\d+\s+F\.?\s*\d*d\s+\d+$',
        # Federal Supplement: 123 F.Supp. 456, 123 F Supp 456, 123 F. Supp. 456
        r'^\d+\s+F\.?\s*Supp\.?\s+\d+$',
        # Federal Appendix: 123 F.App'x 456, 123 F Appx 456
        r'^\d+\s+F\.?\s*App[\'\.]?x\s+\d+$',
        # State citations with series: 123 Cal. 3d 456, 123 N.Y. 2d 456
        r'^\d+\s+[A-Za-z]+\.?\s+\d+d\s+\d+$',
        # State citations without series: 123 Cal. 456, 123 N.Y. 456
        r'^\d+\s+[A-Za-z]+\.?\s+\d+$',
    ]
    
    for pattern in patterns:
        if re.match(pattern, citation_clean):
            return True
    
    # If no pattern matches, raise validation error
    raise ValidationError(
        f"Invalid citation format: '{citation}'. "
        f"Expected formats include: '576 U.S. 644', '123 F.3d 456', '123 F.Supp. 456', "
        f"'123 Cal. 3d 456', or similar standard legal citation formats."
    )


def validate_docket_number(docket_number: str) -> bool:
    """
    Validate docket number format.
    
    Supports common docket number formats including:
    - SCOTUS: "21-123"
    - Federal District: "1:21-cv-12345" or "1-21-cv-12345"
    - Federal Circuit: "21-1234"
    - State: "CR-21-12345" or "CV-2021-12345"
    - Simple numeric: "12345"
    
    Args:
        docket_number: Docket number to validate
    
    Returns:
        True if valid, raises ValidationError if invalid
    
    Raises:
        ValidationError: If docket number format is invalid
    """
    if not docket_number:
        raise ValidationError("Docket number cannot be empty")
    
    # Comprehensive docket number patterns
    patterns = [
        # SCOTUS: 21-123
        r'^\d+-\d+$',
        # Federal District with colon: 1:21-cv-12345, 1:21-cr-12345, 1:21-mc-12345
        r'^\d+:\d+-(cv|cr|mc|md|mi|mj|bk|ap)-\d+$',
        # Federal District with hyphen: 1-21-cv-12345, 1-21-cr-12345
        r'^\d+-\d+-(cv|cr|mc|md|mi|mj|bk|ap)-\d+$',
        # Federal Circuit: 21-1234, 21-12345
        r'^\d+-\d{4,}$',
        # State with prefix: CR-21-12345, CV-2021-12345, CA-21-12345
        r'^[A-Z]{1,4}-\d{2,4}-\d+$',
        # Simple numeric: 12345
        r'^\d+$',
    ]
    
    for pattern in patterns:
        if re.match(pattern, docket_number, re.IGNORECASE):
            return True
    
    # If no pattern matches, raise validation error
    raise ValidationError(
        f"Invalid docket number format: '{docket_number}'. "
        f"Expected formats include: '21-123' (SCOTUS), '1:21-cv-12345' (Federal), "
        f"'CR-21-12345' (State), or numeric formats."
    )


def validate_court_id(court_id: str) -> bool:
    """
    Validate court ID format.
    
    Args:
        court_id: Court ID to validate
    
    Returns:
        True if valid, raises ValidationError if invalid
    
    Raises:
        ValidationError: If court ID format is invalid
    """
    if not court_id:
        raise ValidationError("Court ID cannot be empty")
    
    # Court IDs are typically lowercase alphanumeric with hyphens
    pattern = r'^[a-z0-9-]+$'
    if not re.match(pattern, court_id):
        raise ValidationError(f"Invalid court ID format: {court_id}")
    
    return True


def validate_api_token(token: str) -> bool:
    """
    Validate API token format.
    
    Args:
        token: API token to validate
    
    Returns:
        True if valid, raises ValidationError if invalid
    
    Raises:
        ValidationError: If token format is invalid
    """
    if not token:
        raise ValidationError("API token cannot be empty")
    
    # API tokens are typically alphanumeric
    if len(token) < 10:
        raise ValidationError("API token appears too short")
    
    return True


def validate_id(id_value: Union[int, str]) -> bool:
    """
    Validate ID value.
    
    Args:
        id_value: ID to validate
    
    Returns:
        True if valid, raises ValidationError if invalid
    
    Raises:
        ValidationError: If ID format is invalid
    """
    if id_value is None:
        raise ValidationError("ID cannot be None")
    
    if isinstance(id_value, int):
        if id_value <= 0:
            raise ValidationError("ID must be a positive integer")
    elif isinstance(id_value, str):
        if not id_value.strip():
            raise ValidationError("ID string cannot be empty")
        try:
            int_val = int(id_value)
            if int_val <= 0:
                raise ValidationError("ID must be a positive integer")
        except ValueError:
            raise ValidationError(f"Invalid ID format: {id_value}")
    else:
        raise ValidationError(f"ID must be int or str, got {type(id_value)}")
    
    return True


def validate_url(url: str) -> bool:
    """
    Validate URL format.
    
    Args:
        url: URL to validate
    
    Returns:
        True if valid, raises ValidationError if invalid
    
    Raises:
        ValidationError: If URL format is invalid
    """
    if not url:
        raise ValidationError("URL cannot be empty")
    
    # Basic URL pattern
    pattern = r'^https?://[^\s/$.?#].[^\s]*$'
    if not re.match(pattern, url):
        raise ValidationError(f"Invalid URL format: {url}")
    
    return True


def validate_required_field(value: any, field_name: str) -> bool:
    """
    Validate that a required field is not None or empty.
    
    Args:
        value: Value to validate
        field_name: Name of the field for error messages
    
    Returns:
        True if valid, raises ValidationError if invalid
    
    Raises:
        ValidationError: If field is None or empty
    """
    if value is None:
        raise ValidationError(f"{field_name} is required")
    
    if isinstance(value, str) and not value.strip():
        raise ValidationError(f"{field_name} cannot be empty")
    
    return True 

File: php/src/CourtListener/Utils/Validators.php
<?php

namespace CourtListener\Utils;

/**
 * Validation utilities for the CourtListener SDK
 * 
 * @package CourtListener\Utils
 */
class Validators
{
    /**
     * Validate a date string
     *
     * @param string|null $date Date string to validate
     * @param string $format Expected date format
     * @return bool
     */
    public static function validateDate(?string $date, string $format = 'Y-m-d'): bool
    {
        if ($date === null || $date === '') {
            return false;
        }

        $d = \DateTime::createFromFormat($format, $date);
        return $d && $d->format($format) === $date;
    }

    /**
     * Validate a citation string
     *
     * @param string|null $citation Citation to validate
     * @return bool
     */
    public static function validateCitation(?string $citation): bool
    {
        if ($citation === null || $citation === '') {
            return false;
        }

        // Basic citation validation patterns
        $patterns = [
            '/^\d+\s+U\.S\.\s+\d+$/',  // SCOTUS: "123 U.S. 456"
            '/^\d+\s+F\.\s*\d+$/',     // Federal Reporter: "123 F. 456"
            '/^\d+\s+F\.\s*Supp\.\s*\d+$/', // Federal Supplement: "123 F. Supp. 456"
            '/^\d+\s+[A-Za-z]+\.?\s*[A-Za-z]*\.?\s*\d+$/',   // State citations: "123 Cal. 456" or "123 N.Y. 456"
        ];

        foreach ($patterns as $pattern) {
            if (preg_match($pattern, $citation)) {
                return true;
            }
        }

        return false;
    }

    /**
     * Validate a docket number
     *
     * @param string|null $docketNumber Docket number to validate
     * @return bool
     */
    public static function validateDocketNumber(?string $docketNumber): bool
    {
        if ($docketNumber === null || $docketNumber === '') {
            return false;
        }

        // Common docket number patterns
        $patterns = [
            '/^\d+-\d+-\w+-\d+$/',       // 1-23-cv-456
            '/^\d+:\d+-\w+-\d+$/',       // 1:23-cv-456
            '/^\d+-\w+-\d+$/',           // 1-cv-456
            '/^\d+$/',                   // Simple number
            '/^\d+-\d+$/',               // 1-23
        ];

        foreach ($patterns as $pattern) {
            if (preg_match($pattern, $docketNumber)) {
                return true;
            }
        }

        return false;
    }

    /**
     * Validate a court ID
     *
     * @param string|null $courtId Court ID to validate
     * @return bool
     */
    public static function validateCourtId(?string $courtId): bool
    {
        if ($courtId === null || $courtId === '') {
            return false;
        }

        // Court ID patterns (alphanumeric with hyphens)
        return preg_match('/^[a-zA-Z0-9\-_]+$/', $courtId) === 1;
    }

    /**
     * Validate an API token
     *
     * @param string|null $token API token to validate
     * @return bool
     */
    public static function validateApiToken(?string $token): bool
    {
        if ($token === null || $token === '') {
            return false;
        }

        // API token should be at least 20 characters
        return strlen($token) >= 20;
    }

    /**
     * Validate an ID (positive integer)
     *
     * @param mixed $id ID to validate
     * @return bool
     */
    public static function validateId($id): bool
    {
        if (is_int($id)) {
            return $id > 0;
        }

        if (is_string($id)) {
            return ctype_digit($id) && (int)$id > 0;
        }

        return false;
    }

    /**
     * Validate a URL
     *
     * @param string|null $url URL to validate
     * @return bool
     */
    public static function validateUrl(?string $url): bool
    {
        if ($url === null || $url === '') {
            return false;
        }

        // Only allow HTTP and HTTPS URLs
        if (!filter_var($url, FILTER_VALIDATE_URL)) {
            return false;
        }

        $parsed = parse_url($url);
        return isset($parsed['scheme']) && in_array($parsed['scheme'], ['http', 'https']);
    }

    /**
     * Validate a required field
     *
     * @param mixed $value Value to validate
     * @param string $fieldName Field name for error messages
     * @return bool
     */
    public static function validateRequired($value, string $fieldName = 'field'): bool
    {
        if ($value === null) {
            return false;
        }

        if (is_string($value)) {
            return trim($value) !== '';
        }

        if (is_array($value)) {
            return !empty($value);
        }

        return true;
    }

    /**
     * Validate email address
     *
     * @param string|null $email Email to validate
     * @return bool
     */
    public static function validateEmail(?string $email): bool
    {
        if ($email === null || $email === '') {
            return false;
        }

        return filter_var($email, FILTER_VALIDATE_EMAIL) !== false;
    }

    /**
     * Validate phone number
     *
     * @param string|null $phone Phone to validate
     * @return bool
     */
    public static function validatePhone(?string $phone): bool
    {
        if ($phone === null || $phone === '') {
            return false;
        }

        // Remove common phone number formatting
        $cleaned = preg_replace('/[^\d]/', '', $phone);
        
        // US phone numbers should be 10 or 11 digits
        return strlen($cleaned) >= 10 && strlen($cleaned) <= 11;
    }

    /**
     * Validate a case name
     *
     * @param string|null $caseName Case name to validate
     * @return bool
     */
    public static function validateCaseName(?string $caseName): bool
    {
        if ($caseName === null || $caseName === '') {
            return false;
        }

        // Case name should be at least 3 characters and contain letters
        return strlen(trim($caseName)) >= 3 && preg_match('/[a-zA-Z]/', $caseName);
    }

    /**
     * Validate a judge name
     *
     * @param string|null $judgeName Judge name to validate
     * @return bool
     */
    public static function validateJudgeName(?string $judgeName): bool
    {
        if ($judgeName === null || $judgeName === '') {
            return false;
        }

        // Judge name should be at least 5 characters and contain letters
        return strlen(trim($judgeName)) >= 5 && preg_match('/[a-zA-Z]/', $judgeName);
    }

    /**
     * Validate a search query
     *
     * @param string|null $query Search query to validate
     * @return bool
     */
    public static function validateSearchQuery(?string $query): bool
    {
        if ($query === null || $query === '') {
            return false;
        }

        // Search query should be at least 2 characters
        return strlen(trim($query)) >= 2;
    }

    /**
     * Validate pagination parameters
     *
     * @param array $params Pagination parameters
     * @return bool
     */
    public static function validatePaginationParams(array $params): bool
    {
        if (isset($params['page']) && (!is_int($params['page']) || $params['page'] < 1)) {
            return false;
        }

        if (isset($params['per_page']) && (!is_int($params['per_page']) || $params['per_page'] < 1 || $params['per_page'] > 100)) {
            return false;
        }

        return true;
    }

    /**
     * Validate date range parameters
     *
     * @param array $params Date range parameters
     * @return bool
     */
    public static function validateDateRangeParams(array $params): bool
    {
        $startDate = $params['date_created__gte'] ?? $params['date_filed__gte'] ?? null;
        $endDate = $params['date_created__lte'] ?? $params['date_filed__lte'] ?? null;

        if ($startDate && !self::validateDate($startDate)) {
            return false;
        }

        if ($endDate && !self::validateDate($endDate)) {
            return false;
        }

        if ($startDate && $endDate) {
            $start = \DateTime::createFromFormat('Y-m-d', $startDate);
            $end = \DateTime::createFromFormat('Y-m-d', $endDate);
            
            if ($start && $end && $start > $end) {
                return false;
            }
        }

        return true;
    }

    /**
     * Validate filter parameters
     *
     * @param array $filters Filter parameters
     * @return bool
     */
    public static function validateFilters(array $filters): bool
    {
        // Validate pagination
        if (!self::validatePaginationParams($filters)) {
            return false;
        }

        // Validate date range
        if (!self::validateDateRangeParams($filters)) {
            return false;
        }

        // Validate search query if present
        if (isset($filters['q']) && !self::validateSearchQuery($filters['q'])) {
            return false;
        }

        return true;
    }
}
